### PR TITLE
BUG: fix json segfaults

### LINKED ISSUE #10778: df.to_json segfaults with categorical column types
This code from the Categorical tutorial:

```python
df = pd.DataFrame({"A":["a","b","c","a"]})
df["B"] = df["A"].astype('category')
```

Will segfault when `df.to_json` is called.  Tested on:

```bash
Python 3.4.3 |Anaconda 2.3.0 (x86_64)| (default, Mar  6 2015, 12:07:41)
[GCC 4.2.1 (Apple Inc. build 5577)] on darwin
$ conda list  | grep pandas
pandas                    0.16.2               np19py34_0
```

### LINKED ISSUE #11299: pandas.io.tests.test_json.test_ujson.NumpyJSONTests testOdArray segfaults under python2.7/3.5 in 0.17.0
```
testOdArray (pandas.io.tests.test_json.test_ujson.NumpyJSONTests) ... Segmentation fault (core dumped)
```
```python
In [2]: pd.show_versions()

INSTALLED VERSIONS
------------------
commit: None
python: 2.7.10.final.0
python-bits: 64
OS: Linux
OS-release: 4.1.8-lh
machine: x86_64
processor: Intel(R) Core(TM) i7-2600 CPU @ 3.40GHz
byteorder: little
LC_ALL: None
LANG: en_GB.utf8

pandas: 0.17.0
nose: 1.3.7
pip: 7.1.2
setuptools: 18.4
Cython: 0.23.3
numpy: 1.9.3
scipy: 0.16.0
statsmodels: 0.6.1
IPython: 4.0.0
sphinx: 1.3.1
patsy: 0.4.0
dateutil: 2.4.2
pytz: 2015.6
blosc: None
bottleneck: 1.0.0
tables: 3.2.2
numexpr: 2.4.4
matplotlib: 1.4.3
openpyxl: 2.2.6
xlrd: 0.9.4
xlwt: 1.0.0
xlsxwriter: 0.7.4
lxml: 3.4.4
bs4: 4.4.1
html5lib: 0.9999999
httplib2: 0.9.2
apiclient: 1.3.1
sqlalchemy: 1.0.8
pymysql: 0.6.6.None
psycopg2: 2.6.1 (dt dec pq3 ext lo64)
```
Abrt report: https://gist.github.com/jlec/6282f0dae0586d6749af

### LINKED ISSUE #11473: to_json segfaults with timezone-aware datetimes
```
>>> import pandas as pd; import pytz; import datetime; faulthandler
>>> faulthandler.enable()
>>> pd.DataFrame.from_records([{'ts': datetime.datetime(2015,1,1, tzinfo=pytz.UTC)}]).to_json()

Fatal Python error: Segmentation fault

Current thread 0x00007fff73afd300 (most recent call first):
  File "<snip>lib/python2.7/site-packages/pandas/io/json.py", line 78 in write
  File "<snip>lib/python2.7/site-packages/pandas/io/json.py", line 35 in to_json
  File "<snip>lib/python2.7/site-packages/pandas/core/generic.py", line 893 in to_json
  File "<stdin>", line 1 in <module>
Segmentation fault: 11
```

### PR DESCRIPTION
**Pull request by Komnomnomnom**
closes #11473
closes #10778
closes #11299
- [x] tests added / passed
- [x] vbench / asv ok
- [x] windows tests pass
- [x] valgrind clean
- [x] passes `git diff upstream/master | flake8 --diff`
- [x] whatsnew entry

This fixes several potential segfaults in the json code:
- time objects were freed and potentially garbage collected before their data had been read #11473. Was also segfaulting if an exception was raised during conversion (e.g. when using dateutil timezones).
- 0d arrays were not being handled corectly (needed `goto INVALID`) #11299
- all blocks were assumed to be ndarrays #10778 

Fixing #10778 means non-ndarray blocks are now supported (although I think at present `category` is the only one?).

Not tested on windows yet. Seeing some unrelated travis failures on my fork (msgpack), is that normal?

### CODE REVIEW

**Comment by jreback** [comment]
non-ndarray values are the ``extension`` types

- ``categorical``
- ``datetime w/tz``
- ``sparse`` (which are actually a ndarray sub-class); not a big deal ATM.

**Comment by jreback** [comment]
for tz

```
In [5]: df = DataFrame({'A' : pd.date_range('20130101',periods=3,tz='US/Eastern'), 'B' : pd.date_range('20130101',periods=3)})

In [6]: df
Out[6]: 
                          A          B
0 2013-01-01 00:00:00-05:00 2013-01-01
1 2013-01-02 00:00:00-05:00 2013-01-02
2 2013-01-03 00:00:00-05:00 2013-01-03

In [7]: df.dtypes
Out[7]: 
A    datetime64[ns, US/Eastern]
B                datetime64[ns]
dtype: object

In [8]: df.to_json()
/Users/jreback/miniconda/envs/coursebuild/bin/python.app: line 3: 49679 Segmentation fault: 11  /Users/jreback/miniconda/envs/coursebuild/python.app/Contents/MacOS/python "$@"
```

**Comment by Komnomnomnom** [comment]
welp, ok I'll take another look.

**Comment by jreback** [comment]
@Komnomnomnom thanks!

**Comment by jreback** [comment]
pls update when you can

**Comment by Komnomnomnom** [comment]
The json code assumed in several places that the `.values` attribute would always return an `ndarray`, which of course is not the case for any of the extension types. It then makes extensive use of the numpy C-API to iterate over the assumed `ndarray`, which is a terrible idea for non-ndarray objects (or even ndarray subclasses in the case of sparse).

I added a helper function to handle this a bit better. It initially attempts to use the `.values` attribute and falls back to `get_values()` if necessary. If `get_values()` fails an exception is raised. Subclasses of `ndarray` are not accepted. Seems to work ok now for `sparse` and `categorical`, but it could be better as some coercion / copying is needed.

`datetime w/tz` no longer segfaults but it is not JSON serializable as both `values` and `get_values()` do not return an `ndarray`. I'm inclined to think this is a bug?

```python
In [35]: df
Out[35]: 
                          A
0 2013-01-01 00:00:00-05:00
1 2013-01-02 00:00:00-05:00
2 2013-01-03 00:00:00-05:00

In [36]: df.values
Out[36]: 
DatetimeIndex(['2013-01-01 00:00:00-05:00', '2013-01-02 00:00:00-05:00',
               '2013-01-03 00:00:00-05:00'],
              dtype='datetime64[ns, US/Eastern]', freq='D')

In [37]: df.get_values()
Out[37]: 
DatetimeIndex(['2013-01-01 00:00:00-05:00', '2013-01-02 00:00:00-05:00',
               '2013-01-03 00:00:00-05:00'],
              dtype='datetime64[ns, US/Eastern]', freq='D')

In [38]: df._data.blocks[0].values
Out[38]: 
DatetimeIndex(['2013-01-01 00:00:00-05:00', '2013-01-02 00:00:00-05:00',
               '2013-01-03 00:00:00-05:00'],
              dtype='datetime64[ns, US/Eastern]', freq='D')

In [39]: df._data.blocks[0].get_values()
Out[39]: 
DatetimeIndex(['2013-01-01 00:00:00-05:00', '2013-01-02 00:00:00-05:00',
               '2013-01-03 00:00:00-05:00'],
              dtype='datetime64[ns, US/Eastern]', freq='D')

In [40]: type(df.values)
Out[40]: pandas.tseries.index.DatetimeIndex

In [41]: df.dtypes
Out[41]: 
A    datetime64[ns, US/Eastern]
dtype: object

In [42]: type(df._data.blocks[0])
Out[42]: pandas.core.internals.DatetimeTZBlock

In [43]: df._data.blocks[0].dtype
Out[43]: datetime64[ns, US/Eastern]
```

**Comment by jreback** [comment]
@Komnomnomnom no that is all as expected. Internally the ``DatetimeTZBlock`` holds a ``DatetimeIndex`` rather than a ndarray (though it is duck-like a ndarray).

How do you think we should serialize this?

In general converting to a ndarray is not a great thing as these are now UTC times (don't be fooled as to the repr, they have been converted).
```
In [44]: np.array(df._data.blocks[1].values)
Out[44]: 
array(['2013-01-01T05:00:00.000000000', '2013-01-02T05:00:00.000000000',
       '2013-01-03T05:00:00.000000000'], dtype='datetime64[ns]')
```

converting to a list will preserve everything
```
In [42]: list(df._data.blocks[1].values)
Out[42]: 
[Timestamp('2013-01-01 00:00:00-0500', tz='US/Eastern', offset='D'),
 Timestamp('2013-01-02 00:00:00-0500', tz='US/Eastern', offset='D'),
 Timestamp('2013-01-03 00:00:00-0500', tz='US/Eastern', offset='D')]
```

This converts to string, not sure how useful this is to you (we use this for csv writing).
```
In [54]: df._data.blocks[1].values.to_native_types()
Out[54]: 
array(['2013-01-01 00:00:00-05:00', '2013-01-02 00:00:00-05:00',
       '2013-01-03 00:00:00-05:00'], dtype=object)
```

This is probably the most general thing you can do (and will coerce all things correctly). So if you find a block that doesn' thave ``.values`` as something you recoginize you can just do this.
```
In [57]: df._data.blocks[1].get_values(dtype=object)
Out[57]: 
array([Timestamp('2013-01-01 00:00:00-0500', tz='US/Eastern'),
       Timestamp('2013-01-02 00:00:00-0500', tz='US/Eastern'),
       Timestamp('2013-01-03 00:00:00-0500', tz='US/Eastern')], dtype=object)
```

**Comment by jreback** [comment]
lmk when you are ready for review

**Comment by jreback** [comment]
@Komnomnomnom hows this coming

**Comment by Komnomnomnom** [comment]
@jreback should have some time this evening

**Comment by Komnomnomnom** [comment]
(on UK time)

**Comment by jreback** [comment]
@Komnomnomnom great thanks!

**Comment by Komnomnomnom** [comment]
@jreback TZ block is supported now. Note atm the JSON handler **always** converts to UTC when encoding and this seems to be the case for any time object. It would be nice if it honoured TZ timestamps  (see #12997) but I've added some regression tests to ensure this behaviour remains consistent in the meantime.

So the behaviour now is to repeatedly call `.values` on an object until a concrete `ndarray` is returned with a fall-back to `get_values()` if that doesn't work. So the resulting calls end up being:

 - `DatetimeTZBlock` -> `tzblock.values.values`
 - `SparseBlock` -> `spblock.values.values`
 - `CategorocalBlock` -> `catblock.get_values()`

I just want to give it a run through vbench, otherwise I'm happy with it. 

Let me know if you have any comments. Am I missing something regarding the TZ behaviour?

**Comment by jreback** [comment]
@Komnomnomnom no problem with coercing to UTC, that's what we do anyone when transforming to any non-tz supported.

note that this is numpy 1.11, under older ones you would get a differernt DISPLAY (but the actual values are correct).
```
In [1]: s = Series(pd.date_range('20130101',periods=3,tz='US/Eastern'))

In [2]: s
Out[2]: 
0   2013-01-01 00:00:00-05:00
1   2013-01-02 00:00:00-05:00
2   2013-01-03 00:00:00-05:00
dtype: datetime64[ns, US/Eastern]

In [3]: s.values
Out[3]: 
array(['2013-01-01T05:00:00.000000000', '2013-01-02T05:00:00.000000000',
       '2013-01-03T05:00:00.000000000'], dtype='datetime64[ns]')
```

File: doc/source/whatsnew/v0.18.1.txt [hunk 1]
```diff
@@ -321,6 +322,9 @@ Deprecations
 
 
 
+- Potential segfault in ``DataFrame.to_json`` when serialising ``datetime.time`` (:issue:`11473`).
```
Comment (jreback) [inline doc/source/whatsnew/v0.18.1.txt]: add the issue for the datetime w/tz fixes?

File: pandas/io/tests/test_json/test_pandas.py [hunk 2]
```diff
@@ -821,6 +821,96 @@ def my_handler_raises(obj):
...
+
+        df["B"] = df["A"].astype('category')
+        self.assertEqual(expected, df.to_json())
+
+        s = df["A"]
+        sc = df["B"]
+        self.assertEqual(s.to_json(), sc.to_json())
+
+    def test_datetime_tz(self):
+        tz_range = pd.date_range('20130101', periods=3, tz='US/Eastern')
```
Comment (jreback) [inline pandas/io/tests/test_json/test_pandas.py]: ideally put the issue number associated (for those tests which you added and its appropriate)

File: pandas/io/tests/test_json/test_pandas.py [hunk 3]
```diff
@@ -821,6 +821,96 @@ def my_handler_raises(obj):
...
+    def test_tz_is_utc(self):
+        exp = '"2013-01-10T05:00:00.000Z"'
+
+        ts = Timestamp('2013-01-10 05:00:00Z')
+        self.assertEqual(exp, pd.json.dumps(ts, iso_dates=True))
+        dt = ts.to_datetime()
+        self.assertEqual(exp, pd.json.dumps(dt, iso_dates=True))
+
+        ts = Timestamp('2013-01-10 00:00:00', tz='US/Eastern')
+        self.assertEqual(exp, pd.json.dumps(ts, iso_dates=True))
```
Comment (jreback) [inline pandas/io/tests/test_json/test_pandas.py]: is ``iso_dates`` a new kw?

File: doc/source/whatsnew/v0.18.1.txt (hunk 1 above)
Comment (Komnomnomnom) [inline doc/source/whatsnew/v0.18.1.txt]: `DatetimeTZBlock` was covered by #10778 mentioned [here](https://github.com/pydata/pandas/pull/12802/files#diff-95743d156b5649e3cce3e70750d36248R397). (#11473 is a separate issue related to  `datetime` objects)

File: pandas/io/tests/test_json/test_pandas.py (hunk 2 above)
Comment (Komnomnomnom) [inline pandas/io/tests/test_json/test_pandas.py]: ty, done

File: pandas/io/tests/test_json/test_pandas.py (hunk 3 above)
Comment (Komnomnomnom) [inline pandas/io/tests/test_json/test_pandas.py]: it's a kw on `pd.json.dumps`. `date_format` is the corresponding keyword on `pd.to_json`

**Comment by Komnomnomnom** [comment]
@jreback is there CI support for windows? If not is there any chance you could run the JSON tests on your windows box?

**Comment by jreback** [comment]
yes, you can sign up for http://www.appveyor.com/.

we run this on master, but you can run on a personal (free) accct.

I will run now (on my local vm).

**Comment by jreback** [comment]
@Komnomnomnom passes!

**Comment by jreback** [comment]
side issue. for conformity I was going to rename:

```pandas\io\tests\test_json\`` to ``pandas\io\tests\json\`` (also have to make a tiny change in ``setup.py``. Want to do that here?

**Comment by jreback** [comment]
or I can do it in a followup.

**Comment by Komnomnomnom** [comment]
Great!

np I'll rename it now

**Comment by jreback** [comment]
@Komnomnomnom this looks fine. ping me when your travis run is green (no need to wait for master)

**Comment by jreback** [comment]
@Komnomnomnom 

also move the ``test_json_norm.py`` in there as well.

**Comment by jreback** [comment]
@Komnomnomnom looks good. ping when green.

**Comment by Komnomnomnom** [comment]
Some unrelated msgpack and pickle failures (during version comparison?), not sure what's going on there. Otherwise clean:

https://travis-ci.org/Komnomnomnom/pandas/builds/125890385

**Comment by jreback** [comment]
@Komnomnomnom thanks. yeah those happen when it doesn't get the tag information, not really sure why it doesnt' always work.

**Comment by jreback** [comment]
thanks for the fixes!

**Pull request closed without merging**